packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1.ts

```diff
@@ -67,15 +67,19 @@ function getDeps<TExpression>(
 
 function getDep<TExpression>(depObj: AstObject<R3DeclareDependencyMetadata, TExpression>):
     R3DependencyMetadata {
+  const attribute = depObj.has('attribute') ? depObj.getOpaque('attribute') : null;
+  // Exactly one of `token` or `attribute` will appear on `depObj`.
+  // We want the error to indicate that `token` is missing if neither appear.
+  const token = attribute ?? depObj.getOpaque('token');
   const dep: R3DependencyMetadata = {
-    token: depObj.getOpaque('token'),
+    token,
     host: depObj.has('host') && depObj.getBoolean('host'),
     optional: depObj.has('optional') && depObj.getBoolean('optional'),
     self: depObj.has('self') && depObj.getBoolean('self'),
     skipSelf: depObj.has('skipSelf') && depObj.getBoolean('skipSelf'),
   };
-  if (depObj.has('attribute')) {
-    dep.attribute = depObj.getOpaque('attribute')
-  };
+  if (attribute) {
```

Comment:
```suggestion
  if (attribute !== null) {
```

But actually, `R3DependencyMetadata.attribute` should be required and nullable instead?

Thinking about it more, this is deviating from what `attribute` ought to be: a type representation of the attribute name. The linker now has to set it as a marker, but its value is not otherwise used. And the opaque `attribute` expression in the partial declaration may not be a valid type representation. For the `dynamicAttrName` case, the AOT compiler uses `token: Wrapped(Call(receiver="dynamicAttrName")), attribute: Wrapped(KeywordType(unknown))` and the linker would now provide `token: Wrapped(Call("dynamicAttrName")), attribute: Wrapped(Call(receiver="dynamicAttrName"))`. So having `R3ResolvedDependencyType` internally (but not in the partial declaration) may still be desirable.

Reply:
I agree that it is a bit obfuscated, but adding back in a flag is redundant when the existence of the `attribute` property is sufficient to indicate that this needs to be specially injected.

I am going to rename `attribute` to `attributeNameType` in the internal `R3DependencyMetadata` interface, since that better describes its meaning and purpose. Then in outside of full AOT (JIT mode and linking) we will pass a dummy value (the literal string `"unknown"`) into this property for attributes and I will document this in the code.

Reply:
Further in the `R3DeclareDependencyMetadata` we should only have the `token` and a boolean flag `attribute` since in linking we don't care about the attribute name type. This fits better with the other flags `host`, `optional`, etc.